topology/probes/k8s/cache.go

```diff
@@ -0,0 +1,199 @@
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"time"
+
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology/graph"
+
+	api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ResourceHandler is used to map Kubernetes resources to objets in the graph
+type ResourceHandler interface {
+	Map(obj interface{}) (graph.Identifier, graph.Metadata, error)
+	Dump(obj interface{}) string
+	IsTopLevel() bool
+}
+
+type DefaultResourceHandler struct {
+}
+
+func (h *DefaultResourceHandler) IsTopLevel() bool {
+	return false
+}
+
+// ResourceCache describes a cache for a specific kind of Kubernetes resource.
+// It is in charge of listening Kubernetes events and creating the
+// according resource in the graph with the informations returned by
+// the associated resource handler
+type ResourceCache struct {
+	*graph.GraphEventHandler
+	cache          cache.Indexer
+	controller     cache.Controller
+	stopController chan (struct{})
+	handler        ResourceHandler
+	graph          *graph.Graph
+}
+
+func (c *ResourceCache) list() []interface{} {
+	return c.cache.List()
+}
+
+func (c *ResourceCache) getByKey(namespace, name string) interface{} {
+	key := ""
+	if len(namespace) > 0 {
+		key = namespace + "/"
+	}
+	key += name
+	if obj, found, _ := c.cache.GetByKey(key); found {
+		return obj
+	}
+	return nil
+}
+
+func (c *ResourceCache) getByNode(node *graph.Node) interface{} {
+	namespace, _ := node.GetFieldString("Namespace")
+	name, _ := node.GetFieldString("Name")
+	if name == "" {
+		return nil
+	}
+	return c.getByKey(namespace, name)
+}
+
+func (c *ResourceCache) getByNamespace(namespace string) []interface{} {
+	if namespace == api.NamespaceAll {
+		return c.list()
+	}
+
+	objects, _ := c.cache.ByIndex("namespace", namespace)
+	return objects
+}
+
+func (c *ResourceCache) getBySelector(g *graph.Graph, namespace string, selector *metav1.LabelSelector) []metav1.Object {
+	if objects := c.getByNamespace(namespace); len(objects) > 0 {
+		return filterObjectsBySelector(objects, selector)
+	}
+	return nil
+}
+
+// OnAdd is called when a new Kubernetes resource has been created
+func (c *ResourceCache) OnAdd(obj interface{}) {
+	c.graph.Lock()
+	defer c.graph.Unlock()
+
+	id, metadata, err := c.handler.Map(obj)
+	if err != nil {
+		logging.GetLogger().Debugf("Failed to map %s: %s", c.handler.Dump(obj), err)
+	} else {
+		node := c.graph.NewNode(id, metadata, "")
+		c.NotifyEvent(graph.NodeAdded, node)
+		logging.GetLogger().Debugf("Added %s", c.handler.Dump(obj))
+	}
+}
+
+// OnUpdate is called when a Kubernetes resource has been updated
+func (c *ResourceCache) OnUpdate(oldObj, newObj interface{}) {
+	c.graph.Lock()
+	defer c.graph.Unlock()
+
+	if id, metadata, err := c.handler.Map(newObj); err == nil {
```

Comment:
the `OnAdd` emits a log if Map() failed ... either both OnUpdate/OnAdd emit this log ... or neither. I would favor not emitting a log as if Map failed this is more of some internal errors (which should be debugged / fixed) rather than due to some rare external input

Reply:
Remove the error returned by Map (was never used)